File: src/main/java/frc/team6223/robot/Robot.java
package frc.team6223.robot;


import edu.wpi.first.wpilibj.Preferences;
import edu.wpi.first.wpilibj.command.Command;
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser;
import frc.team6223.arsenalFramework.drive.ArsenalDrive;
import frc.team6223.arsenalFramework.hardware.ArsenalNavXMicro;
import frc.team6223.arsenalFramework.hardware.ArsenalRobot;
import frc.team6223.arsenalFramework.hardware.motor.ArsenalTalon;
import frc.team6223.arsenalFramework.operator.ArsenalOperatorInterface;
import frc.team6223.arsenalFramework.software.commands.ArcadeDriveController;
import frc.team6223.arsenalFramework.software.commands.ForceMovementController;
import frc.team6223.arsenalFramework.software.units.Time;
import frc.team6223.arsenalFramework.software.units.TimeUnits;
import frc.team6223.robot.subsystem.Climber;


public class Robot extends ArsenalRobot {

    private ArsenalDrive drive;

    private Climber climberSubsystem = new Climber(
        new ArsenalTalon(3, false, false, false),
        new ArsenalTalon(7, false, false, false),
        new ArsenalTalon(4, false, false, false)
    );

    public Robot() {
        super(DEFAULT_PERIOD, 0.05);
    }

    @Override
    protected SendableChooser<Command> injectAutonomousChooser() {
        SendableChooser<Command> sendableChooser = new SendableChooser<>();
        sendableChooser.addDefault(
          "Move forward for 2 sec",
          new ForceMovementController(
            drive,
            new Time(4.0, TimeUnits.SECONDS),
            0.5, 0.5
          )
        );
        return sendableChooser;
    }

    @Override
    protected void injectSendableChoosers() {
        // no-op
    }

    @Override
    protected void allocateSubsystems(Preferences prefs) {
        drive = new ArsenalDrive(
          new ArsenalNavXMicro(),
          new ArsenalTalon(1),
          new ArsenalTalon(2)
        );
    }

    @Override
    protected ArsenalOperatorInterface allocateOperatorInterface(Preferences prefs) {
        return new OI(climberSubsystem);
    }

    @Override
    protected void setTeleoperatedCommand() {
          new ArcadeDriveController(drive, this.operatorInterface.getPrimaryJoystick(), 1.0).start();
    }

    public void autonomousInit() {
        super.autonomousInit();
        this.drive.resetEncoders();
    }

    public void teleopInit() {
        super.teleopInit();
        this.drive.resetEncoders();
    }

    @Override
    public void dashboardPeriodic() {
        this.drive.dashboardPeriodic();
    }
}

File: src/main/kotlin/frc/team6223/robot/Robot.kt
package frc.team6223.robot

import edu.wpi.first.wpilibj.*
import edu.wpi.first.wpilibj.command.Command
import edu.wpi.first.wpilibj.smartdashboard.SendableChooser
import frc.team6223.arsenalFramework.drive.ArsenalDrive
import frc.team6223.arsenalFramework.hardware.ArsenalNavXMicro
import frc.team6223.arsenalFramework.hardware.ArsenalRobot
import frc.team6223.arsenalFramework.hardware.motor.ArsenalTalon
import frc.team6223.arsenalFramework.hardware.motor.ArsenalVictor
import frc.team6223.arsenalFramework.operator.ArsenalOperatorInterface
import frc.team6223.arsenalFramework.software.commands.MoveDriveTrainCommand
import frc.team6223.arsenalFramework.software.controllers.ArcadeDriveController
import frc.team6223.arsenalFramework.software.controllers.ForceMovementController
import frc.team6223.arsenalFramework.software.controllers.NoMovementController
import frc.team6223.arsenalFramework.software.units.*
import frc.team6223.robot.auto.AutoUtilities
import frc.team6223.robot.conf.LEFT_DRIVE_CONTROLLER
import frc.team6223.robot.conf.RIGHT_DRIVE_CONTROLLER
import frc.team6223.robot.subsystem.Climber

class Robot: ArsenalRobot(TimedRobot.DEFAULT_PERIOD, 0.05) {

    private val climberSubsystem = Climber(
            ArsenalTalon(3, false, false, false),
            ArsenalTalon(7, false, false, false)
                    .also { it.addFollower(ArsenalVictor(9, false)) },
            ArsenalTalon(4, false, false, false)
                    .also { it.addFollower(ArsenalVictor(8, false)) }
    )
    private lateinit var driveSubsystem: ArsenalDrive
    //private val pdpSubsystem = PDP(PDP_CAN_ID)
    private val robotSideChooser = AutoUtilities.generateSendableChooser()

    override fun injectAutonomousChooser(): SendableChooser<Command> {
        val sendableChooser = SendableChooser<Command>()
        sendableChooser.addDefault(
                "Move forward for 2 sec",
                MoveDriveTrainCommand(
                        ForceMovementController(
                                Time(4.0, TimeUnits.SECONDS),
                                0.5, 0.5
                        ),
                        driveSubsystem
                )
        )
        /*sendableChooser.addDefault(
                        "Move 15ft? using Motion Profiling",
                        MoveDriveTrainCommand(
                                MotionProfileController(
                                        arrayOf(
                                                Waypoint(0.0, 0.0, 0.0),
                                                Waypoint(15.0, 0.0, 0.0)
                                        ),
                                        Trajectory.Config(
                                                Trajectory.FitMethod.HERMITE_QUINTIC,
                                                Trajectory.Config.SAMPLES_HIGH,
                                                0.02,
                                                1.0,
                                                1.0,
                                                60.0),
                                        Velocity(Distance(6.54, DistanceUnits.FEET), Time(1.0, TimeUnits.SECONDS))),
                                driveSubsystem)
                )
                */
        return sendableChooser
    }

    override fun allocateSubsystems(preferences: Preferences) {
        driveSubsystem = ArsenalDrive(
                NoMovementController(),
                ArsenalNavXMicro(),
                ArsenalTalon(LEFT_DRIVE_CONTROLLER, true, true, false, false),
                ArsenalTalon(RIGHT_DRIVE_CONTROLLER, true, true, true, false)
        )
    }

    override fun injectSendableChoosers() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun allocateOperatorInterface(preferences: Preferences): ArsenalOperatorInterface {
        return OI(climberSubsystem)
    }

    override fun setTeleoperatedCommand() {
        MoveDriveTrainCommand(ArcadeDriveController(operatorInterface.primaryJoystick), driveSubsystem).start()
    }

    override fun autonomousInit() {
        super.autonomousInit()
        this.driveSubsystem.resetEncoders()
        //this.driveSubsystem.zeroYaw()
    }

    override fun teleopInit() {
        super.teleopInit()
        this.driveSubsystem.resetEncoders()
    }

    override fun dashboardPeriodic() {
        this.driveSubsystem.dashboardPeriodic()
    }
}
